### PR TITLE
Move stack into CCM on F3/F4 targets

### PR DESCRIPTION
**Pull request by digitalentity**
Closes https://github.com/iNavFlight/inav/issues/787

Tested on SPRacingF3 and REVO

### CODE REVIEW

File: src/main/target/link/stm32_flash_f303_128k.ld [hunk 1]
```diff
@@ -14,7 +14,8 @@ MEMORY
 {
   FLASH  (rx)     : ORIGIN = 0x08000000, LENGTH = 126K /* last 2kb used for config storage */
   RAM    (xrw)    : ORIGIN = 0x20000000, LENGTH = 40K
+  CCM    (xrw)    : ORIGIN = 0x10000000, LENGTH = 8K
   MEMORY_B1 (rx)  : ORIGIN = 0x60000000, LENGTH = 0K
 }
 
-INCLUDE "stm32_flash.ld"
+INCLUDE "stm32_flash_ccm.ld"
```
Comment (martinbudden) [inline src/main/target/link/stm32_flash_f303_128k.ld]: Do we need to add a new linker script? Can't we just check if CCM is defined in the existing script?

Comment (digitalentity) [reply src/main/target/link/stm32_flash_f303_128k.ld]: as far as I know, ld doesn't support conditions, however i'm not an ld expert

Comment (martinbudden) [reply src/main/target/link/stm32_flash_f303_128k.ld]: I'm not an expert on this either. @ledvinap can you help here?

Comment (ledvinap) [reply src/main/target/link/stm32_flash_f303_128k.ld]: It's possible to use C preprocessor on linker file first. All goodies from C preprocessor are available then (memory sizes from Makefile, target, ...).

https://github.com/ledvinap/cleanflight/blob/local/Makefile#L890
It is possbile without grep, but `-ggdb3` must be removed from `$(CFLAGS)` for this gcc invocation

There was discussion about this, version with minimal target-specific files was used. Personally I prefer preprocessor (everything is in one place), but that is matter of taste

File: src/main/startup/startup_stm32f411xe.s [hunk 2]
```diff
@@ -72,6 +72,9 @@ defined in linker script */
   .weak  Reset_Handler
   .type  Reset_Handler, %function
 Reset_Handler: 
+  // Enable CCM
+
```
Comment (martinbudden) [inline src/main/startup/startup_stm32f411xe.s:82]: Looks like you haven't inserted the Enable CCM code here.

File: src/main/target/link/stm32_flash_f303_128k.ld (hunk 1 above)
Comment (martinbudden) [reply src/main/target/link/stm32_flash_f303_128k.ld]: I'm certainly in favour of using the preprocessor and keeping a single  `stm32_flash.ld` file. One less `.ld` file to maintain.

Comment (martinbudden) [reply src/main/target/link/stm32_flash_f303_128k.ld]: @ledvinap Could we use the `REGION_ALIAS` command? Ie
`REGION_ALIAS("STACKRAM", RAM)`
or
`REGION_ALIAS("STACKRAM", CCM)`

according to the `.ld` file, and then setup the stack in `STACKRAM`?

**Comment by martinbudden** [comment]
@digitalentity , I've done a "for discussion" PR in betaflight https://github.com/betaflight/betaflight/pull/1670 , using `REGION_ALIAS`. Let's see what people say.

**Comment by digitalentity** [comment]
Won't work, but I have an idea how to do it in a better way than it is now.
Will require some testing though.

**Comment by martinbudden** [comment]
Note the one thing this does break is stack watermarking, since the linker doesn't fill the CCM stack with 0xa5.

**Comment by digitalentity** [comment]
It actually doesn't break watermarking. We fill the stack RAM with 0xa5 during startup. 

In fact, linker directive to fill the stack is ignored - we are not loading ELF file, so flags to fill segments with pre-defined values are ignored. We fill BSS with zeros during startup as well.

**Comment by digitalentity** [comment]
Tested on SPRACINGF3 & REVO. Working as expected, stack is at CCM top, `STACK_CHECK` also works.

**Comment by martinbudden** [comment]
BTW, you forgot to remove the "Don't merge" label.

I like this kind of PR - free speed with no downsides.

**Comment by digitalentity** [comment]
Actually, CCM doesn't give us much performance benefit if used only for stack. We probably need to concentrate on https://github.com/iNavFlight/inav/pull/619 and move bus transfer to background.

**Comment by martinbudden** [comment]
I know it doesn't give much performance benefit (I've done some measurements and they confirm it). But still, it's FREE speed.

**Comment by digitalentity** [comment]
And also free RAM. Now we can have i.e. more waypoints.

**Pull request merged**